### ISSUE TITLE
Cannot build with -DNDEBUG

### ISSUE THREAD

**Issue by unknownzerx** [scylladb/seastar#116]
If I compile with the flag `DNDEBUG`, it will complain
```
/root/seastar/core/future.hh:953:13: error: control reaches end of non-void function [-Werror=return-type]
```
near the `assert`
```c++
future<T...> finally(Func&& func) noexcept {
    // ...
    if (!f_res.failed()) {
        return std::move(result);
    } else {
        if (!result.failed()) {
            return make_exception_future<T...>(f_res.get_exception());
        } else {
            // ...
            assert(0 && "we should not be here");
        }
    }
    // ...
}
```
(-Wall and -Werr are defined in SEASTAR_CFLAGS)

The solution could be return or throw some dummy object.

**Comment by nyh** [comment]
You are right: By defining the NDEBUG macro, all assert()s can be compiled away, so we must write all our assertions with this in mind: assertions must never have side-effects (e.g., never do "assert(do_something() == 0)") and as you noted, assert(0) **cannot** be assumed to never return. 

When we really want to abort in case of a bug, I think we should use **abort()**, not assert(0), nor throw or return a dummy object. This is not only place in the code which uses assert(0), and they should probably all be fixed, but you're right that the place you found in the header file is the most important, because application code uses this header file, and this application code might choose to compile with NDEBUG, even if Seastar itself was compiled without this flag.

**Comment by xemul** [comment]
Seems to build fine now (see https://github.com/scylladb/seastar/pull/3554#issuecomment-5103560585)
